### PR TITLE
Grid Accept Dataframe

### PR DESCRIPTION
**Pull request by Kully**
For issue: https://github.com/plotly/plotly.py/issues/665

Just a first pass. Haven't added tests yet - want to make sure my approach is good enough.

### CODE REVIEW

**Comment by Kully** [comment]
@theengineear Mind taking a look when you have a minute?

File: plotly/grid_objs/grid_objs.py [hunk 1]
```diff
@@ -9,7 +9,9 @@
 
 from requests.compat import json as _json
 
-from plotly import exceptions, utils
+from plotly import exceptions, optional_imports, utils
+
+pd = optional_imports.get_module('pandas')
```
Comment (theengineear) [inline plotly/grid_objs/grid_objs.py]: 👌

File: plotly/grid_objs/grid_objs.py [hunk 2]
```diff
@@ -148,7 +150,21 @@ def __init__(self, columns_or_json, fid=None):
         ```
         """
         # TODO: verify that columns are actually columns
-        if isinstance(columns_or_json, dict):
+        if pd and isinstance(columns_or_json, pd.DataFrame):
+            column_names = [name for name in columns_or_json.columns]
```
Comment (theengineear) [inline plotly/grid_objs/grid_objs.py]: 🐄 if you *need* this to be a list, why not just `list(columns_or_json.columns)`? If you *don't* need this to be a list, why not just leave it as a pandas array? `column_names = columns_or_json.columns`?

File: plotly/grid_objs/grid_objs.py [hunk 3]
```diff
@@ -148,7 +150,21 @@ def __init__(self, columns_or_json, fid=None):
...
+            duplicate_name = utils.get_first_duplicate(column_names)
```
Comment (theengineear) [inline plotly/grid_objs/grid_objs.py]: ah, here we go, can you not just do `utils.get_first_duplicate(columns_or_json.columns)`?

File: plotly/grid_objs/grid_objs.py [hunk 4]
```diff
@@ -148,7 +150,21 @@ def __init__(self, columns_or_json, fid=None):
...
+            if duplicate_name:
+                err = exceptions.NON_UNIQUE_COLUMN_MESSAGE.format(duplicate_name)
+                raise exceptions.InputError(err)
+
+            # create columns from dataframe
+            all_columns = []
+            for name in column_names:
```
Comment (theengineear) [inline plotly/grid_objs/grid_objs.py]: Same here, embrace the :duck:. `for name in columns_or_json.columns:`.

File: plotly/grid_objs/grid_objs.py [hunk 5]
```diff
@@ -148,7 +150,21 @@ def __init__(self, columns_or_json, fid=None):
...
+                all_columns.append(Column(columns_or_json[name].tolist(), name))
+            self._columns = list(all_columns)
```
Comment (theengineear) [inline plotly/grid_objs/grid_objs.py]: Why `list` this? You've already instantiated a new list above `all_columns = []`. I don't see why an additional shallow copy is necessary?

File: plotly/grid_objs/grid_objs.py (hunk 4 above)
Comment (Kully) [inline plotly/grid_objs/grid_objs.py]: Don't know why I'm afraid of Ducks...

File: plotly/tests/test_optional/test_grid/test_grid.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+
+
+class Test_Dataframe_to_Grid(TestCase):
+
+    # Test duplicate columns
+    def test_duplicate_columns(self):
+        df = pd.DataFrame([[1, 'a'],
+                           [2, 'b']], columns=['col_1', 'col_1'])
+
+        NON_UNIQUE_COLUMN_MESSAGE = (
```
Comment (theengineear) [inline plotly/tests/test_optional/test_grid/test_grid.py]: 🐄 no real need for this to be a `CONSTANT`, imo. no biggie though.

Comment (Kully) [inline plotly/tests/test_optional/test_grid/test_grid.py]: Do you mean rename it to something like `pattern` or `message` rather than `ALLCAPSCONSTANT`?

Comment (theengineear) [inline plotly/tests/test_optional/test_grid/test_grid.py]: `expected_message` or something is what I meant.

File: plotly/tests/test_optional/test_grid/test_grid.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+from unittest import TestCase
+
+from plotly.exceptions import InputError
+from plotly.grid_objs import Grid
+
+import pandas as pd
+
+
+class Test_Dataframe_to_Grid(TestCase):
```
Comment (Kully) [inline plotly/tests/test_optional/test_grid/test_grid.py]: Two things:
1) What does 🐍 mean?
2) Why would DRY be relevant here? What am I repeating?

I'm changing it to your suggested class name, just curious

Comment (theengineear) [inline plotly/tests/test_optional/test_grid/test_grid.py]: Sorry, I can see why that's confusing ;__;.

I was talking about camelCase vs snake_case (hence the 🐍 and 🐫).

**Pull request merged**